Add block navigator movers

talldan opened the pull request:
## Description
Closes #11408

Adds block navigator movers to the navigation block.

This PR refactors the Block Navigator to use follow a TreeGrid as described in https://w3c.github.io/aria-practices/examples/treegrid/treegrid-1.html?cell=force, as suggested by feedback on this PR. The navigator can be navigated using arrow keys.

Some descriptions have been added to the buttons for selecting blocks and the appender for screenreader users using `aria-describedby`.

This is quite a large PR, so there are a couple of PRs extracted from this:
- https://github.com/WordPress/gutenberg/pull/19799 introduces RovingTabIndex and TreeGrid components
- https://github.com/WordPress/gutenberg/pull/22122 allows the individual mover buttons to be imported, which is needed in this PR so that focus can be managed separately on the buttons.

## How has this been tested?
1. Add a navigation menu block (ensure the feature is enabled as an experiment first)
2. Add some menu items to the navigation menu
3. Click the block navigator button on the toolbar
4. Try reordering the items using the movers

<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

## Screenshots <!-- if applicable -->
![navigator-movers](https://user-images.githubusercontent.com/677833/81152486-9088ef80-8fb4-11ea-8dfd-a3083aefa72b.gif)

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
New feature (non-breaking change which adds functionality)
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->

draganescu commented:
So far this works great and the code is perfect.

talldan commented:
I think this one is open for reviews now. 

I think it needs some a11y feedback and design feedback. 

There are some issues I've spotted that I'll continue to look into:
- When clicking the up mover button, focus is retained, but clicking the down mover button causes a focus loss
- When selected item is focused the background turns white
- Select block button uses the new display name, but screenreader text for mover buttons still refer to item by its block type
- I think it would be good to read what's read by a screenreader for the select button. e.g. the select button should probably read 'Select [Block Type]' instead of just the block type.

mtias commented on packages/block-editor/src/components/block-navigation/list.js:
```diff
@@ -44,38 +45,48 @@ export default function BlockNavigationList( {
 								blocks={ block.innerBlocks }
 								selectedBlockClientId={ selectedBlockClientId }
 								selectBlock={ selectBlock }
+								showMovers={ showMovers }
```
Maybe "showBlockMovers" would be a clearer prop.

mtias commented on packages/block-editor/src/components/block-navigation/list.js:
```diff
@@ -44,38 +45,48 @@ export default function BlockNavigationList( {
+						>
 							<Button
-								className={ classnames( 'editor-block-navigation__item-button block-editor-block-navigation__item-button', {
-									'is-selected': isSelected,
```
Losing this selected state means after moving a block I can't tell which block I just moved.

talldan replied:
There is selected state still, it was moved from the button to the wrapping div, so that entire row for the block shows as selected. 

I think the issue is that the block doesn't become selected when clicking a mover.

youknowriad reviewed (commented):
Can we benefit from all these Block Navigation improvements in the global block navigation popover as well?

mtias commented:
> Can we benefit from all these Block Navigation improvements in the global block navigation popover as well?

The movers is the only one I'd be comfortable enabling in the top level one, but we need some improvements first — selection should be clearer, ideally also support the motion effect we have for moving blocks in the canvas.

mtias commented:
> Allow moving blocks in and out of nesting (follow-up PR?)
> Drag and drop (follow-up PR?)
> Animate moving of blocks (follow-up PR?)

Follow up PR seems fine. We can look at enabling the movers in the top-level navigator once these are worked out.

tellthemachines reviewed (dismissed):
I did a bit of testing on NVDA/Firefox and VoiceOver/Safari. Agree that screen reader output for mover buttons is not very informative right now, would be better to read out something like "move Sample page menu item".

I'm afraid the issue with focus being lost when clicking the down mover is even worse using NVDA/Firefox as focus moves to the document body, even with the modal still open (yikes!). On Firefox with no screen reader, focus can also be moved out of the modal but not every time. Here's an example screenshot of Firefox on Mac with focus moved behind the modal and the menu item still visually selected:
<img width="605" alt="Screen Shot 2019-10-28 at 2 14 21 pm" src="https://user-images.githubusercontent.com/8096000/67651063-7af69580-f993-11e9-9f86-458ac54acca0.png">

Another issue I'm noticing with the Navigator modal is that clicking on any arrow key closes it. This was already happening before this PR so it's not related to these changes but, because the navigator list now has a role of `tree`, VoiceOver is announcing it as a table, which is expected to be navigable with arrow keys. (Using the VO keys + arrow keys still works, but just using plain arrows should also work in a table scenario.) Still, it might be easier to handle this in a separate PR.

talldan commented:
Thanks for testing and reviewing @tellthemachines. There's definitely a bit subtlety to getting these movers to work in the right way.

It's interesting in your screenshot that focus has moved back to the mover outside the popover. Coincidence? I'll do some more debugging/testing to establish.

tellthemachines commented:
> It's interesting in your screenshot that focus has moved back to the mover outside the popover. Coincidence? I'll do some more debugging/testing to establish.

Initially it moves to the block inserter, then if you tab once it goes to the mover. I forgot to mention that bit, sorry.

talldan commented:
I've managed to solve the focus loss issue when clicking the move down button. The code is less than ideal though, it requires a bit of tidy up. 

I think I'll need to dive into refactoring how the `BlockMover` component handles visibility. Currently it's quite opinionated and would have been developed for the one use case, being displayed in the main block list. It doesn't work quite so well with the different selection model in the Block Navigator.

talldan commented on packages/block-editor/src/components/block-navigation/list.js:
```diff
@@ -44,38 +45,48 @@ export default function BlockNavigationList( {
+								{ blockDisplayName }
 								{ isSelected && <span className="screen-reader-text">{ __( '(selected block)' ) }</span> }
 							</Button>
+							{ hasMovers && ( <BlockMover clientIds={ [ block.clientId ] } /> ) }
```
I've switched to using the `isHidden` prop. 

It seems a shame that JS has to be used for this instead of CSS, but in practice it seems like the reason is to replicate `:focus-within`—ensuring the whole mover component continues to be displayed when any of the buttons within the row have focus.

talldan commented on packages/block-editor/src/components/block-navigation/style.scss:
```diff
@@ -16,6 +16,44 @@ $tree-item-height: 36px;
+		}
+	}
+
+	&:focus-within {
```
Now removed as per the comment above.

talldan commented:
This should be ready for a re-review.

The only bit of the feedback from @tellthemachines I haven't tackled is that the aria text for the mover buttons doesn't refer to the block in the same way as the selection button. There's a separate PR for that as it's not an insignificant refactor - https://github.com/WordPress/gutenberg/pull/18132.

draganescu commented:
@talldan there seems to be a kind of weird behaviour pictured below:

![movers-weirdness](https://user-images.githubusercontent.com/107534/68474367-dba48e80-022d-11ea-834b-f2176a614881.gif)

1st the mover stays where I clicked - I think this is complicated, if it moves with the item I have to move my mouse to chase it to the top, however this feels more natural
2nd, I get two sets of movers when moving items down b/c the show on hover. Maybe add hover intent behaviour to the list items?

mtias commented:
To improve the clarity of the interaction, see self plug :) https://matiasventura.com/post/using-motion-to-express-change/

We should add this behaviour to the navigator as well.

talldan commented:
@draganescu Yep, that's a curiosity that I noticed too. It actually worked that way for the standard block movers back before the block animations and before the movers were only shown on selected blocks:

![movers](https://user-images.githubusercontent.com/677833/68574296-832af680-04a4-11ea-867b-43ac5d018b04.gif)

I think it would have been less apparent because blocks are generally different sizes, so less chance the moved block will end up in the same position.

I think I'll add some animations as @mtias suggests and then re-assess whether we need to do anything about this.

talldan commented:
@mtias They're now animated!

noisysocks commented on packages/block-editor/src/components/block-navigation/list.js:
```diff
@@ -40,59 +44,29 @@ function getBlockDisplayName( blockType, attributes ) {
-				const blockType = getBlockType( block.name );
-				const isSelected = block.clientId === selectedBlockClientId;
-
+		<ul className="editor-block-navigation__list block-editor-block-navigation__list" role={ isTreeRoot ? 'tree' : 'group' }>
```
Is it correct that, if `isTreeRoot` is false, the `<ul>` will have `role="group"` and the `<li>`s will have `role="treeitem"`?

talldan replied:
Yeah, the inner `ul`s should have be groups, if you scroll down a bit here there's some example markup:
https://www.w3.org/TR/wai-aria-practices/examples/treeview/treeview-2/treeview-2a.html

noisysocks commented on packages/block-editor/src/components/block-navigation/list.js:
```diff
@@ -101,6 +75,88 @@ export default function BlockNavigationList( {
+	);
+}
+
+export default function BlockNavigationList( props ) {
```
Why not simply `export default NavigationList;`?

talldan replied:
I've recombined NavigationList and BlockNavigationList, so this should be resolved now.

noisysocks commented on packages/components/src/modal/frame.js:
```diff
@@ -70,6 +70,10 @@ class ModalFrame extends Component {
 			this.handleEscapeKeyDown( event );
 		}
+
+		if ( this.props.onKeyDown ) {
```
Should we add this new prop to the modal's readme and update the changelog?

talldan replied:
Yeah, I'm not sure about this prop. It is needed, but I might make it experiemental so that there's the option of removing it in the future. We seem to have this issue regularly where we need to stop propagation of keyboard events from popovers and modals, and it'd be good to come up with a better pattern.

talldan replied:
Changed it now so that this is handled internally modal frame using a prop `__experimentalPreventKeyboardEventPropagation`.

noisysocks commented on packages/block-editor/src/components/block-navigation/list.js:
```diff
@@ -40,59 +44,29 @@ function getBlockDisplayName( blockType, attributes ) {
-	parentBlockClientId,
-} ) {
-	const shouldShowAppender = showAppender && !! parentBlockClientId;
+function NavigationList( { blocks, selectBlock, selectedBlockClientId, showAppender, showBlockMovers, showNestedBlocks, parentBlockClientId } ) {
```
nit: I liked having each prop on a new line 🙂

talldan replied:
They're back on a new line in `BlockNavigationList` and `NavigationBlock` has fewer props.

noisysocks reviewed (dismissed):
Nothing blocking here 🙂 — this all works really well in my testing and is a great improvement. I love love the refactoring of `BlockNavigationList` into `NavigationList` and `NavigationBlock`.

tellthemachines commented:
> The other bit I'm unsure about is the best way to make the menu navigable using arrow keys. The tricky part is that each row now contains multiple buttons. At the moment up/down arrow keys cycle through in the same way tab does.

What we have here might be closer to a [treegrid](https://www.w3.org/TR/wai-aria-practices/#treegrid). 
The expected keyboard interaction for these widgets is [clearly specified](https://www.w3.org/TR/wai-aria-practices/#keyboard-interaction-22) so I guess we should follow the spec in that respect. It's quite close to what we have now, but we need to change a few things: 
* tab key should not go through the tree, only enter or exit it
* left/right arrows can be used to navigate horizontally, up/down only vertically ([this example](https://www.w3.org/TR/wai-aria-practices/examples/treegrid/treegrid-1.html) is good)
* up/down arrows shouldn't cycle through but stop when hitting the top/bottom.

I have two concerns regarding using these patterns (tree or treegrid) for the navigator:  
* the spec states that they should be expandable/collapsible. Is that a plan for this component?
* it is a rare enough pattern in web apps that keyboard-only users may not know how to navigate it (should we add some discreet instructions?)

talldan commented:
I've implemented the treegrid arrow key navigation as described by @tellthemachines. I need to move some of the code around as at the moment it's all implemented in the block library in the folder for the navigation menu.

The roving tab index is a little harder to implement and I haven't quite tackled that yet.

noisysocks commented on packages/block-library/src/navigation/block-navigation-list.js:
```diff
@@ -8,6 +13,100 @@ import {
+	};
+
+	return (
+		// eslint-disable-next-line jsx-a11y/no-static-element-interactions
```
A comment explaining why we're disabling this rule would be good.

noisysocks commented on packages/block-library/src/navigation/block-navigation-list.js:
```diff
@@ -8,6 +13,100 @@ import {
+	} );
+}
+
+function NavigableTreeGrid( { children } ) {
```
A short comment here explaining why we need this component and what pressing left/right/up/down does would be helpful.

noisysocks commented on packages/block-library/src/navigation/block-navigation-list.js:
```diff
@@ -8,6 +13,100 @@ import {
+				nextRowIndex = Math.min( currentRowIndex + 1, rows.length - 1 );
+			}
+
+			// Navigation is either at the top or bottom of the grid. Do nothing.
```
Why do we cycle back to the beginning or end when navigating horizontally (`LEFT`, `RIGHT`) but not vertically?

noisysocks commented on packages/block-library/src/navigation/block-navigation-list.js:
```diff
@@ -30,12 +129,15 @@ export default function BlockNavigationList( { clientId } ) {
-			showNestedBlocks
-			showAppender
-		/>
+		<NavigableTreeGrid>
```
This is mind bogglingly nuanced stuff—great job! 🎉

noisysocks commented on packages/block-library/src/navigation/block-navigation-list.js:
```diff
@@ -30,12 +129,15 @@ export default function BlockNavigationList( { clientId } ) {
-			showAppender
-		/>
+		<NavigableTreeGrid>
+			<__experimentalBlockNavigationList
```
Should `<NavigableTreeGrid>` be inside `<__experimentalBlockNavigationList>`? When would you want to use `<__experimentalBlockNavigationList>` and **not** have the grid functionality?

noisysocks commented on packages/block-library/src/navigation/use-block-navigator.js:
```diff
@@ -42,6 +42,7 @@ export default function useBlockNavigator( clientId ) {
 			onRequestClose={ () => {
 				setIsNavigationListOpen( false );
 			} }
+			__experimentalPreventKeyboardEventPropagation
```
Not a big fan of this prop. Could an alternative be to wrap the modal contents in a container?

```js
<Modal
	...
>
	<div onKeyDown={ ( event ) => {
		// Insert explanatory comment here
		if ( includes( [ LEFT, DOWN, RIGHT, UP, BACKSPACE, ENTER ], event.keyCode ) ) {
			event.stopPropagation();
		}
	} }>
		<BlockNavigationList ... />
	</div>
</Modal>
```

Alternatively, might be worth looking into if `IgnoreNestedEvents` can help us here. This is how we get `BlockListBlock` to ignore events in other parts of the codebase.

```js
<Modal
	...
>
	<IgnoreNestedEvents childHandledEvents={ [ 'onKeyDown' ] }>
		<BlockNavigationList ... />
	</IgnoreNestedEvents>
</Modal>
```

talldan replied:
So the issue is that the modal is closed not only when the keyboard nav happens within the modal contents, but also when the modal UI itself is interacted with (e.g. the close button), so surrounding just the BlockNavigationList isn't sufficient.

I've got a different solution which I've split into a separate PR https://github.com/WordPress/gutenberg/pull/19804

I think a better solution is a more global one. I've added it to this PR, but also split it out into #18014

noisysocks reviewed (commented):
This works really well and I love the grid. It feels really natural to navigate around using the arrow keys 😍

I am not very into `__experimentalPreventKeyboardEventPropagation`—is there an alternative?

mapk reviewed (dismissed):
I just tested this out and it feels right from a design perspective. 👍 

![nav 2020-01-03 16_09_07](https://user-images.githubusercontent.com/617986/71756241-983c5a00-2e43-11ea-9364-83774d49d9b1.gif)

I do have one concern which will most likely be remedied in another PR. It sounds like we're going to [add these improvements to the Top Toolbar ](https://github.com/WordPress/gutenberg/pull/18014#issuecomment-546296733)as well which I think is a good idea. (see #19357 )

Once we do, this block will have 3 different areas in which the same Navigator can be accessed (ie. Top Toolbar Navigator, Block Navigator, Sidebar Navigator). This seems quite redundant. I initially liked the sidebar option because it allowed interaction without covering the block with a popover, but now these three options seem like a bit much. Any thoughts on how we might solve this redundancy? Should we solve it?

![Screen Shot 2020-01-03 at 4 09 41 PM](https://user-images.githubusercontent.com/617986/71756575-675d2480-2e45-11ea-969c-4a07f9f4df09.png)

mapk commented:
Oooh! I just noticed that all three Navigators are being called something different. Can we call them all the same?

### Top Toolbar "Block navigation"

![Screen Shot 2020-01-03 at 4 25 31 PM](https://user-images.githubusercontent.com/617986/71756679-da669b00-2e45-11ea-8b90-c5737592c585.png)

### Block Toolbar Modal "Block Navigator"

![Screen Shot 2020-01-03 at 4 25 52 PM](https://user-images.githubusercontent.com/617986/71756699-f8340000-2e45-11ea-9475-5d23324e128d.png)

### Sidebar "Navigation Structure"

![Screen Shot 2020-01-03 at 4 26 09 PM](https://user-images.githubusercontent.com/617986/71756704-04b85880-2e46-11ea-8e33-2c5ed61d5855.png)

I'm completely open to what we choose here, but would like them all to be consistent. Maybe "Block Navigator"?

WunderBart commented on packages/block-library/src/navigation/block-navigation-list.js:
```diff
@@ -1,3 +1,8 @@
+/**
+ * External dependencies
+ */
+import { includes } from 'lodash';
```
Can we use the [`Array.prototype.includes()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes) here?

Soean replied:
Unfortunately not, because we still support IE 11.

WunderBart replied:
This is a bit confusing to me. It's stated [here](https://github.com/WordPress/gutenberg/tree/master/packages/block-library#installation), that this code will run in [envs that support ES2015+](http://kangax.github.io/compat-table/es6/), which would imply that IE11 users have to use the polyfills anyway?

karmatosed commented:
My design feedback this point isn't so much about having movers here (because that works), as much as iterating as we consider this. Which means there is no reason I don't think this can't be merged for now.

I would also love to see us reduce the duplication of this being in 2 places now.

I also would love to avoid so many +

![image](https://user-images.githubusercontent.com/253067/72107504-6175b080-3329-11ea-82e6-60af797e0bf5.png)

talldan committed:
Add movers for block navigator

Fix doc examples

Make components experimental

Simplify creation of RovingTabIndex provider value

Add storybook snapshots

Add tests for RovingTabIndex, RovingTabIndexItem

Add tests for TreeGrid components

minor variable name change

Format JS

Import TreeGrid components from @wordpress/components

Update styling for block navigator

Try adding descender line styles

Add some further finesse to styling

Extract animated row to fix animation and use change detection that includes a full path so that all children animate

Patch up code after rebase

Use colSpan instead of colspan

Fix row animation when moving a row quickly multiple times

Remove descender lines from the accessibility tree

Use useCallback for callback

Use aria-hidden on purely presentational descender indicator in tree grid

Update example

Fix arrow key navigation when using Voiceover

Only handle arrow key navigation when no modifiers are pressed

Fix regressions in BlockMoverButton caused by rebase. Rename file to just button

Name buttons in a consistent way

Add screenreader description for focused block in navigator

Add aria-describedby for appender button

Fix selected block movers appearing with darker background

Remove writing flow hack

Fix typo

Allow aria-setsize and aria-posinset on role=row in axe e2e tests

Fix translator comment and use Add as the verb instead of Insert

Add aria-label for navigation structure tree grid

Fix typos

Try to fix tests

More typo fixes

Update snapshot to reflect renaming of test description

Update more e2e test classnames

Rename components to more closely reflect `master`

shaunandrews commented:
This seems to work as expected. My only note is that when hovering over the movers, it can be hard to determine what block they're connected too. Perhaps we could add a border (or some other indicator) to the block to help connect it to it's movers:

<img width="323" alt="image" src="https://user-images.githubusercontent.com/191598/74546238-4560c280-4f18-11ea-8159-951e71180670.png">

tellthemachines commented:
I checked this again with VoiceOver+Safari and I'm afraid using the arrow keys is still causing the modal to close altogether. It's fine with just keyboard navigation without a screen reader though.

Also checked with NVDA+Firefox and the arrow navigation works very well, the only problem is there's no indication of the nested structure. So, say my fifth nav item has some children, for the first child it'll announce `row 7` (the top "Navigation" item also counts as a row) and then when we go right to the mover buttons it'll happily tell us that `move up button unavailable block navigation link is at the beginning of the content and can't be moved up row 7`, which is slightly baffling as we have no way of knowing that the item is nested.

MarcoZehe commented:
> Also checked with NVDA+Firefox and the arrow navigation works very well, the only problem is there's no indication of the nested structure. So, say my fifth nav item has some children, for the first child it'll announce `row 7` (the top "Navigation" item also counts as a row) and then when we go right to the mover buttons it'll happily tell us that `move up button unavailable block navigation link is at the beginning of the content and can't be moved up row 7`, which is slightly baffling as we have no way of knowing that the item is nested.

Could applying aria-level to this help? aria-level is 1-based, is used in tree views and tree grids to indicate the nesting of elements.

mtias commented:
@talldan where are we on this?

talldan commented:
@mtias Sorry for the delay in responding, was AFK so missed your comment. This is mostly functional. There are still some issues to solve though. 

The scope of the PR has become much bigger than I expected, which is why it has taken a while. Much of the nav tree has been re-written to conform to a Treegrid (https://w3c.github.io/aria-practices/examples/treegrid/treegrid-1.html), so it needs some thorough review. I split some of that out into a separate PR (https://github.com/WordPress/gutenberg/pull/19799).

Now that it's a treegrid and an HTML table is being used there are a few things that are hard to style, like the descender lines that indicate the hierarchy. I did manage to get that working, but I had to add a fair bit of code to achieve it. I could perhaps do with some fresh eyes on how I've implemented some parts of this.

I haven't been able to implement what @shaunandrews recommends here - https://github.com/WordPress/gutenberg/pull/19799 - which doesn't seem like it should be something so difficult.

mtias commented:
Is the treegrid the best representation for this view? It seems the example leans quite a bit on tabular data.

talldan commented:
It's the one that fit best when we were assessing. There aren't any others that really fit the model of a tree structure with multiple focusable elements in a row.

One option could be to have the mover buttons separate from the tree structure, just use the tree for selecting blocks, with buttons underneath for other actions (moving, duplicating, deleting). That's similar to what photoshop has for duplicating or deleting layers.

That might not fit the design goals, but it'd mean being able to use a TreeView instead, which is pretty much what it is now in terms of markup.

draganescu reviewed (commented):
I tested this and in this state worked as described although it did make Firefox stutter a bit for larger menus.

enriquesanchez commented:
I'm very confused as to why keyboard interaction works great but once I turn VoiceOver on it all breaks. I'm even more confused knowing that the Welcome Guide modal is marked up in a very similar way and focus management is not an issue there. 

I wonder if this is an issue specific to `role="treegrid"`?  I don't have any prior experience with it, but from https://www.w3.org/TR/2017/REC-wai-aria-1.1-20171214/#treegrid :

>To be keyboard accessible, authors SHOULD manage focus of descendants for all instances of this role, as described in Managing Focus.

Perhaps there's more focus management that is needed with this role?

talldan commented:
Hey @enriquesanchez Thanks for taking a look.

> I'm very confused as to why keyboard interaction works great but once I turn VoiceOver on it all breaks. I'm even more confused knowing that the Welcome Guide modal is marked up in a very similar way and focus management is not an issue there.

I'm not sure either, but I do get issues elsewhere, not just in this modal, I made #18537 a while back and I still see this, even though I'm on a new computer now.

Not a lot of other people have mentioned it, so either not many are testing or it's something to do with my setup.

enriquesanchez commented:
>Not a lot of other people have mentioned it, so either not many are testing or it's something to do with my setup.

@talldan I honestly had never paid much attention to this, but I just tested in several places and I'm seeing the same problem that you mention. Though it only happens with VoiceOver, I'm not seeing any issues with NVDA.

I did some searching online but couldn't find anything specific to this problem. It is very weird.

talldan commented:
I've brought this up to date with `master` again. It was quite outdated, so hopefully I haven't broken anything in the process.

I notice the animation when moving a row (especially using the keyboard and moving multiple times in quick succession) is really janky, so will try to fix that.

adamziel commented:
@talldan any change of splitting it into a few smaller PRs? 44 files changed is a bit hard to review (unless it's just a matter of rebasing?)

talldan commented:
Hey @adamziel. Yeah, this did turn into quite a big PR I've made a couple of smaller PRs from this one:
- Introduce TreeGrid and RovingTabIndex components (https://github.com/WordPress/gutenberg/pull/19799)
- Extract Block Mover Buttons (https://github.com/WordPress/gutenberg/pull/22122)

talldan commented:
It'd be great to get some reviews on this again. 

As a user facing feature, this seems to be working well. There's some code that I'm not completely happy with:
- Using a TreeGrid meant a big refactor, I would've preferred to keep this small, but I understand we want to get the user experience right.
- Reworking the visual lines that indicate the tree structure was very complicated in an HTML table. This is handled in the `DescenderLines` component, the previous implementation was a lot cleaner, being almost entirely implemented using CSS, but it relied on nested lists, which isn't possible with a flat TreeGrid/table.

talldan commented on packages/e2e-tests/config/setup-test-framework.js:
```diff
@@ -244,6 +244,11 @@ async function runAxeTestsForBlockEditor() {
 			'.edit-post-layout__metaboxes',
 			// Ignores elements created by TinyMCE.
 			'.mce-container',
+			// These properties were not included in the 1.1 spec
+			// through error, they should be allowed on role="row":
+			// https://github.com/w3c/aria/issues/558
+			'[role="row"][aria-setsize]',
+			'[role="row"][aria-posinset]',
```
This seems like it should allow the e2e tests to ignore the incorrect parts of the spec, but the tests still seem to be failing. 😕

mapk commented:
I tested this in the sidebar. Here's my experience.

1. The movers worked as expected.
2. When I was unable to move an item, the movers did not show – good!
3. However, on one instance a disabled arrow still showed the highlighted border. Should this happen?

![nav](https://user-images.githubusercontent.com/617986/81223320-ae495b00-8f9a-11ea-8a0b-c3f18f074431.gif)

It all looked quite well to me from a design perspective. I can't wait to get these in!

ZebulanStanphill commented:
Could this be implemented for the general Block Navigation, and not just the Navigation block one?

talldan commented:
@mapk I think it does need the border—the disabled button still needs to retain and indicate that it has focus, it works similarly on the main block movers at the moment:
<img width="367" alt="Screenshot 2020-05-07 at 9 38 19 am" src="https://user-images.githubusercontent.com/677833/81245043-9deeba00-9046-11ea-987a-ac03b9485e5e.png">

talldan commented:
@ZebulanStanphill I don't see any issues to enabling it on the main editor interface. My feeling is that the navigation block will be a good place to test it first.

mtias commented:
Yes, let's test it for a release or two in the navigation block to squash all the bugs.

tellthemachines commented on packages/components/src/roving-tab-index/test/item.js:
```diff
@@ -0,0 +1,66 @@
+		expect( renderer.toJSON() ).toMatchSnapshot();
+	} );
+
+	it( 'allows children to bue declared using a child render function as an alternative to `as`', () => {
```
Typo: "to *be* declared".

tellthemachines commented on packages/components/src/tree-grid/README.md:
```diff
@@ -0,0 +1,129 @@
+		<TreeGrid>
+			<TreeGridRow level={ 1 } positionInSet={ 1 } setSize={ 2 }>
+				<TreeGridCell>
+					{ ( props ) => <Button onClick={ onSelect } { ...props }>Select</Button> }
```
Does the element inside the cell have to be a button? I wouldn't expect that to be the typical case for this component. If we were to have just plain text, like in [the wai-aria example](https://www.w3.org/TR/wai-aria-practices/examples/treegrid/treegrid-1.html), where would focus be transferred to? (This is something we can iterate on later though)

talldan replied:
Yeah, that's a good point! I can think of a few ways to accomplish this, and it shouldn't be too difficult as `TreeGridCell` has all the information it needs.

tellthemachines commented on packages/components/src/tree-grid/index.js:
```diff
@@ -0,0 +1,157 @@
+		<RovingTabIndexContainer>
+			{ /* Disable reason: A treegrid is implemented using a table element. */ }
+			{ /* eslint-disable-next-line jsx-a11y/no-noninteractive-element-to-interactive-role */ }
+			<table { ...props } role="treegrid" onKeyDown={ onKeyDown }>
```
Might be good to add in an `aria-label` here that can be passed a short description of what the treegrid is about.

tellthemachines commented on packages/components/src/tree-grid/test/cell.js:
```diff
@@ -0,0 +1,54 @@
+) );
+
+describe( 'TreeGridCell', () => {
+	it( 'requires TreeGrid to be declared as a parent component somewhere in the component heirarchy', () => {
```
Typo: "hierarchy".

tellthemachines reviewed (approved):
This is working much better on VoiceOver, and is working well on NVDA too, apart from one issue: when entering the `treegrid`, NVDA should go into focus mode automatically, but it's not happening. In browse mode, we are not able to use the arrow keys to navigate the `treegrid`, so users will have to manually switch to focus mode to be able to interact with it. This is a bit annoying because [the wai-aria example](https://www.w3.org/TR/wai-aria-practices/examples/treegrid/treegrid-1.html) correctly switches to focus mode, and the markup used seems to be identical in everything to this one 😞 . @diegohaz any ideas on this?

Approving, because we should get this in and iterate! It would be good to fix the above issue before we extend this functionality to the general block navigator though.

diegohaz commented:
@tellthemachines It seems that it happens because the first active element when you tab into the treegrid is a button, not a `role="gridcell"` or `role="row"`. I think this can be solved by always focusing the row when entering the treegrid, like in the WAI-ARIA example.

diegohaz commented:
Hmm, after doing some more tests, it looks like NVDA messes up with focus/browse randomly when you focus on elements inside a tree grid that aren't rows or grid cells.

It happens even on the WAI-ARIA example. As soon as I focus the link, it switches to browse mode, and I'm no longer able to navigate through the tree grid using arrow keys.

The same happens on `role="grid"` elements. Seems like a bug on NVDA?

![Read description above](https://user-images.githubusercontent.com/3068563/81407743-4ae23900-9112-11ea-841c-72f575fd801c.gif)

diegohaz commented:
Related issue on NVDA: https://github.com/nvaccess/nvda/issues/8395

Looks like it’s expected behavior and NVDA users should use especial keys to navigate the grid, or enable focus mode manually.

tellthemachines commented:
Ooooh, great detective work @diegohaz ! I guess there's nothing else to do on our side then.

talldan commented on packages/block-editor/src/components/block-navigation/branch.js:
```diff
@@ -1,59 +1,100 @@
-			</li>
-		);
-	}
+export default function BlockNavigationBranch( props ) {
```
@adamziel I ended up renaming `BlockNavigationBranch` to `BlockNavigationLeaf` and `BlockNavigationList` to `BlockNavigationBranch`, so this diff looks a bit confusing.

I think that's correct though, a branch should have multiple leaves, and the leaf should be the terminating points on a tree.

talldan commented on packages/block-editor/src/components/block-navigation/tree.js:
```diff
@@ -0,0 +1,41 @@
+ *
+ * @param {Object} props
+ */
+export default function BlockNavigationTree( {
```
@adamziel I renamed `BlockNavigationList` to `BlockNavigationTree` to be consistent with the whole tree/branch/leaf thing.

talldan commented on packages/block-editor/src/components/block-navigation/context.js:
```diff
@@ -0,0 +1,11 @@
+ */
+import { createContext, useContext } from '@wordpress/element';
+
+export const BlockNavigationContext = createContext( {
```
@adamziel I moved the context out into its own file and add a `useBlockNavigationContext` helper.

talldan commented on packages/block-editor/src/components/block-navigation/block-contents.js:
```diff
@@ -0,0 +1,185 @@
+					}
+
+					return (
+						<BlockNavigationBlockContentWrapper as="div">
```
@adamziel This main file now contains a lot of the Wrapper and Slot/Fill logic. I ended up renaming a few things.

The main logic I changed is that the wrapper is now rendered around the slot, while in your PR the wrapper is rendered inside the fill:
https://github.com/WordPress/gutenberg/pull/22210/files#diff-2e071f8429cd2c500b830947ffde4239R226-R231

Pretty open to handling that differently, I think some aspects of this may need to be changed to handle adding the menu described in https://github.com/WordPress/gutenberg/issues/22089.

adamziel replied:
@talldan I went for the "wrapper inside od the fill" approach to make something like this possible:

```js
<BlockNavigationBlockFill>
	<BlockNavigationBlockSelectButton
		ref={ ref }
		block={ block }
		onClick={ onClick }
		isSelected={ isSelected }
		position={ position }
		siblingCount={ siblingCount }
		level={ level }
		{ ...props }
	/>
	<EllipsisMenu {/* some props */} />
</BlockNavigationBlockFill>
```

With "wrapper outside of the fill" approach, these slots are only good for something like replacing the label with RichText. You cannot update/replace the button AND add something before/after a button.

talldan replied:
Hmm, this is made difficult because this PR introduces the movers which are rendered after the fill, and then the designs in #22089 shown the dropdown menu after the movers.

I think most blocks will end up having the dropdown menu, so it probably makes sense to add that as an internal part of `BlockNavigation`, and then if we need to add custom items for particular blocks we can try a slot for the menu.

talldan commented:
In case anyone is wondering, the above comments are in relation to resolving merge conflicts between this and the recently merged https://github.com/WordPress/gutenberg/pull/21948

paaljoachim commented:
Awesome that movers are being added (have been added) to the block navigator!
Are there also plans for drag & drop? As it would be so nice to grab a hold of something and just drop it into the place where I need it.

draganescu commented:
There doesn't appear to be any issue for this specifically, @paaljoachim Maybe you can open one, specifically for the block navigator?

paaljoachim commented:
I went ahead and opened a new issue: https://github.com/WordPress/gutenberg/issues/22297
(I added the Navigation and the Navigator, but can of course adjust to only focus on the Block Navigator.)
